core_lib/src/tool/buckettool.cpp

```diff
@@ -169,11 +184,68 @@ void BucketTool::paintBitmap(Layer* layer)
 
     QPoint point = QPoint(qFloor(getLastPoint().x()), qFloor(getLastPoint().y()));
     QRect cameraRect = mScribbleArea->getCameraRect().toRect();
-    BitmapImage::floodFill(targetImage,
-                           cameraRect,
-                           point,
-                           qPremultiply(mEditor->color()->frontColor().rgba()),
-                           properties.tolerance);
+
+    QRgb fillColor = qPremultiply(mEditor->color()->frontColor().rgba());
+    QRgb origColor = fillColor;
+    if (properties.fillMode == 0)
+    {
+        if (qAlpha(fillColor) == 0)
+        {
+            // Filling in overlay mode with a fully transparent color has no
+            // effect, so we can skip it in this case
+            return;
+        }
+    }
```

Comment:
Just a minor thing but this could be simplified to, It's not a must have change though

```suggestion
    if (properties.fillMode == 0 && qAlpha(fillColor) == 0)
    {
            // Filling in overlay mode with a fully transparent color has no
            // effect, so we can skip it in this case
            return;
    }
```

Reply:
This was intentional so the else check won't be run if fillMode == 0 and qAlpha(fillColor) != 0. A premature and useless optimization? Yes absolutely, but it's not really any worse readability-wise so :shrug:

Reply:
I'd argue that nested branches for the sake of premature optimization is unnecessary complexity and does decrease readability, eg. I find it easier to read one branch with logic operators than to consider a nested branch logic but I guess that's highly subjective. It only goes through the branch once, there is no useful optimization to win here, since you're simply checking if the color is alpha based, which is extremely fast.

You're right with only one branch it's not an issue but it is the beginning of unnecessary complexity by branching.

Reply:
I also feel like this might be changed to a switch statement at a later time if we get more modes (such as that erase mode), in which case a nested if statement would make it a bit more natural to notice and make that change. I'm not really defending this choice, it's neither here nor there for me. If you want to change it go ahead.